src/consumer/base_consumer.rs

```diff
@@ -607,8 +714,16 @@ where
     C: ConsumerContext,
 {
     fn drop(&mut self) {
-        trace!("Destroying consumer: {:?}", self.client.native_ptr()); // TODO: fix me (multiple executions ?)
-        unsafe { rdsys::rd_kafka_consumer_close(self.client.native_ptr()) };
+        trace!("Destroying consumer: {:?}", self.client.native_ptr());
+        if self.group_id.is_some() {
+            if let Err(err) = self.close_queue() {
+                error!("Failed to close consumer queue on drop: {}", err);
+            } else {
+                while !self.closed() {
+                    self.poll(Duration::from_millis(100));
+                }
```

Comment:
Is this safe? Could we drop messages here and have the offsets committed to kafka?

Reply:
That's a good question. My understanding is that messages won't be forwarded here anymore. In this particular case we had to switch [rd_kafka_consumer_close](https://github.com/confluentinc/librdkafka/blob/95a542c87c61d2c45b445f91c73dd5442eb04f3c/src/rdkafka.h#L4207-L4212) by [rd_kafka_consumer_close_queue](https://github.com/confluentinc/librdkafka/blob/95a542c87c61d2c45b445f91c73dd5442eb04f3c/src/rdkafka.h#L4228-L4235) because `rd_kafka_consumer_close` will block until the consumer has revoked its assignment by calling `Unassign` or `IncrementalUnassign` on the rebalance handler. This translates on the consumer being blocked forever as there's no rebalance callback registered anymore.

Hence we had to move to `rd_kafka_consumer_close_queue` that allows us to asynchronously close the consumer. In this particular case, by calling `Poll` we can fwd the rebalance event to the rebalance method on the `ConsumerContext` following the API doc:
> The application must poll/serve this queue until rd_kafka_consumer_closed() returns true.

Reply:
Alright, sounds good to me.

Reply:
Is there any chance that the consumer never closed? Because I ran into a situation where my program never exit this while loop.